### ISSUE TITLE
CrossEntropyLoss is negative

### ISSUE THREAD

**Issue by micklexqg** [pytorch/pytorch#2866]
what’s the formula of CrossEntropyLoss used in pytorch?
I used CrossEntropyLoss, but the loss value is negative. generally speaking, it should be positive.
so what’s wrong?

**Comment by soumith** [comment]
The formula is given here: http://pytorch.org/docs/master/nn.html?highlight=crossentropyloss#torch.nn.CrossEntropyLoss

It should not be negative. If it is negative, please give a small test case where we can reproduce this. thank you.

**Comment by micklexqg** [comment]
@soumith , thank you. this is the loss detail, the print frequency is 300.
![image](https://user-images.githubusercontent.com/13776012/30841609-dca24104-a2af-11e7-98d1-50d85c764df9.png)

**Comment by soumith** [comment]
this seems like it might be a sign bug for windows. Are you using the latest package from @peterjc123 ?

**Comment by micklexqg** [comment]
the version is 0.1.12.  there is no pytorch under peterjc123... and I wonder if the  latest package  support windows?

**Comment by soumith** [comment]
please install https://anaconda.org/peterjc123/pytorch

**Comment by micklexqg** [comment]
@soumith , thank you

**Comment by morningmorningwind** [comment]
negative values exist in the latest version of nn.BCELoss().
e.g., criterion = nn.BCELoss()
print(criterion(torch.zeros(5,1),torch.zeros(5,1)))
I got: tensor(1.00000e-12 *
       -1.0001)

my workaround:
criterion(torch.zeros(5,1).clamp(1e-8,1-1e-7),torch.zeros(5,1))

**Comment by soumith** [comment]
@morningmorningwind what is the output of `print(torch.__version__)` and also what OS you are using?

**Comment by morningmorningwind** [comment]
@soumith 
torch v. 0.4.0
python 2.7.14
OS: MacOS High Sierra 10.13.1

**Comment by fmassa** [comment]
@soumith I think this can happen in our current implementation if `x=y=0`, see https://github.com/pytorch/pytorch/blob/master/aten/src/THNN/generic/BCECriterion.c
We would have something like `-log(1 + EPS)`, which is negative.

**Comment by soumith** [comment]
cc: @li-roy can you issue a fix for this.

**Comment by poorvaj-a** [comment]
I am also having the output of the cross entropy function negetive don't know why is that happening.

```
Epoch 1
-------------------------------
loss: -11.750731  [    0/   37]
loss: 2.816855  [    1/   37]
loss: 0.324966  [    2/   37]
loss: 3.002348  [    3/   37]
loss: -0.018466  [    4/   37]
loss: -1.154994  [    5/   37]
loss: -1.036991  [    6/   37]
loss: -3.164487  [    7/   37]
loss: -2.369340  [    8/   37]
loss: -6.293080  [    9/   37]
loss: -5.052002  [   10/   37]
loss: -2.971949  [   11/   37]
loss: -2.884700  [   12/   37]
loss: -2.356856  [   13/   37]
loss: -0.918272  [   14/   37]
loss: -1.373340  [   15/   37]
loss: -2.317873  [   16/   37]
loss: -1.943973  [   17/   37]
loss: -2.920667  [   18/   37]
loss: -3.361804  [   19/   37]
loss: -3.045549  [   20/   37]
loss: -2.117454  [   21/   37]
loss: -2.469415  [   22/   37]
loss: -2.469415  [   23/   37]
loss: -3.595349  [   29/   37]
loss: -3.608938  [   30/   37]
loss: -3.939634  [   31/   37]
loss: -4.156094  [   32/   37]```
